Max time limit for placement engine

soneryaldiz opened the pull request:
This is for #803

854768750 commented:
@soneryaldiz I used `pytest test_circuits.py -k test_cmp --runnightly` but got not failed on my computer. How can I find which test case is failed?

soneryaldiz commented:
> @soneryaldiz I used `pytest test_circuits.py -k test_cmp --runnightly` but got not failed on my computer. How can I find which test case is failed?

I owe you a test and will commit Today.

soneryaldiz commented:
@854768750 , this failure is due to placer finding a solution suboptimal in area:
`AssertionError: Placer found a suboptimal solution: area: 4776408000 target: 4500000000.0 ratio: 1.061424`

I changed the default value to 30 seconds. I added a test which should run for at least 60 seconds: `tests/pdk/finfet/test_placer.py`. 
How are you planning to implement this by the way considering the hierarchy and variants? The intention here is:

1. Force placer to search longer for higher quality solutions,
2. Do not let placer to run forever and finish within an upper bound.

854768750 commented:
> @854768750 , this failure is due to placer finding a solution suboptimal in area: `AssertionError: Placer found a suboptimal solution: area: 4776408000 target: 4500000000.0 ratio: 1.061424`
> 
> I changed the default value to 30 seconds. I added a test which should run for at least 60 seconds: `tests/pdk/finfet/test_placer.py`. How are you planning to implement this by the way considering the hierarchy and variants? The intention here is:
> 
> 1. Force placer to search longer for higher quality solutions,
> 2. Do not let placer to run forever and finish within an upper bound.

I implemented this way. It is still the same result. I used `pytest test_circuits.py -k test_cmp --runnightly` but cannot see error on my computer.

soneryaldiz commented:
> > @854768750 , this failure is due to placer finding a solution suboptimal in area: `AssertionError: Placer found a suboptimal solution: area: 4776408000 target: 4500000000.0 ratio: 1.061424`
> > I changed the default value to 30 seconds. I added a test which should run for at least 60 seconds: `tests/pdk/finfet/test_placer.py`. How are you planning to implement this by the way considering the hierarchy and variants? The intention here is:
> > 
> > 1. Force placer to search longer for higher quality solutions,
> > 2. Do not let placer to run forever and finish within an upper bound.
> 
> I implemented this way. It is still the same result. I used `pytest test_circuits.py -k test_cmp --runnightly` but cannot see error on my computer.

The current implementation seems to achieve # 2 above but I could not see what is forcing placer to search longer? 
See the latest commit, increasing plt to 120 secs does not result in a better solution nor longer search.
"""
cd tests/pdk/finfet/
pytest -v --runnightly test_circuits.py -k cmp_1 -s
"""

srini229 commented:
@soneryaldiz Simulated annealing temperature threshold was the one causing issue. I have fixed it now. Can you try it now? 
Also the way it is currently implemented, the time limit is per hierarchy. If we want it for overall placer, we might have to infer a per hierarchy time limit. One simple way could be allocate time based on number of cells in a hierarchy.

soneryaldiz commented:
@854768750 , @srini229 , I ran few experiments locally (by increasing ALPHA and COUNT_LIMIT hyperparameters) but could not get a higher quality solution. For example, our current selection of ALPHA, T_MIN and T_MAX sets an upper bound of ~5.5k simulated annealing steps. Increasing ALPHA to 0.9975 (doubling the number of steps) did not help at all. 

I will ask help from Steve Tomorrow to isolate a test case that clearly shows smaller area and smaller HPWL on a separate PR for you to explore ideas. Let's keep this PR in draft mode for now.

srini229 commented:
@soneryaldiz I'll add a debug knob to visualize the sequence pairs being used by the placer. This would help triage if the sequence pair generation is the issue.
